crates/turborepo-lib/src/query.rs

```diff
@@ -0,0 +1,411 @@
+            .equal
+            .as_ref()
+            .map(|pair| Self::check_equals(pkg, &pair.field, &pair.value));
+        let not_equal = self
+            .not_equal
+            .as_ref()
+            .map(|pair| !Self::check_equals(pkg, &pair.field, &pair.value));
+
+        let greater_than = self
+            .greater_than
+            .as_ref()
+            .map(|pair| Self::check_greater_than(pkg, &pair.field, &pair.value));
+
+        let less_than = self
+            .less_than
+            .as_ref()
+            .map(|pair| Self::check_greater_than(pkg, &pair.field, &pair.value));
+        let not = self.not.as_ref().map(|predicate| !predicate.check(pkg));
+
+        and.into_iter()
+            .chain(or)
+            .chain(equal)
+            .chain(not_equal)
+            .chain(greater_than)
+            .chain(less_than)
+            .chain(not)
+            .all(|p| p)
+    }
+}
+
+#[Object]
+impl Query {
+    async fn affected_packages(
+        &self,
+        base: Option<String>,
+        head: Option<String>,
+    ) -> Result<Vec<Package>, Error> {
+        let base = base.unwrap_or_else(|| "main".to_string());
+        let head = head.unwrap_or_else(|| "HEAD".to_string());
+        let mut opts = self.run.opts().clone();
+        opts.scope_opts.affected_range = Some((base, head));
+
+        Ok(RunBuilder::calculate_filtered_packages(
+            &self.run.repo_root(),
+            &opts,
+            &self.run.pkg_dep_graph(),
+            &self.run.scm(),
+            &self.run.root_turbo_json(),
+        )?
+        .into_iter()
+        .map(|package| Package {
+            run: self.run.clone(),
+            name: package,
+        })
+        .sorted_by(|a, b| a.name.cmp(&b.name))
+        .collect())
+    }
+    /// Gets a single package by name
+    async fn package(&self, name: String) -> Result<Package, Error> {
+        let name = PackageName::from(name);
+        Ok(Package {
+            run: self.run.clone(),
+            name,
+        })
+    }
+
+    /// Gets a list of packages that match the given filter
+    async fn packages(&self, filter: Option<PackagePredicate>) -> Result<Vec<Package>, Error> {
+        let Some(filter) = filter else {
+            return Ok(self
+                .run
+                .pkg_dep_graph()
+                .packages()
+                .map(|(name, _)| Package {
+                    run: self.run.clone(),
+                    name: name.clone(),
+                })
+                .sorted_by(|a, b| a.name.cmp(&b.name))
+                .collect());
+        };
+
+        Ok(self
+            .run
+            .pkg_dep_graph()
+            .packages()
+            .map(|(name, _)| Package {
+                run: self.run.clone(),
+                name: name.clone(),
+            })
+            .filter(|pkg| filter.check(pkg))
+            .sorted_by(|a, b| a.name.cmp(&b.name))
+            .collect())
+    }
+}
+
+#[Object]
+impl Package {
```

Comment:
would it make sense to add `hash` here?  seems like something we'd want to be able to see/show

Reply:
Packages don't really have a concept of a hash, tasks do. But yeah, eventually there should be task hashes. That will require some reworking of our infrastructure since right now we do things in a direct pipeline fashion, so computing one hash will require computing them all